lib/ancestry/class_methods.rb

```diff
@@ -29,30 +29,25 @@ def orphan_strategy= orphan_strategy
 
     # Arrangement
     def arrange options = {}
-      # Get all nodes ordered by ancestry and start sorting them into an empty hash
-      arrange_nodes self.ancestry_base_class.reorder(options.delete(:order)).where(options)
+      # Get all nodes and start sorting them into an empty hash
+      scope = self.ancestry_base_class
+      if (order = options.delete(:order))
+        scope = scope.order(order)
+      end
+      arrange_nodes scope.where(options)
     end
 
     # Arrange array of nodes into a nested hash of the form
     # {node => children}, where children = {} if the node has no children
+    # If a node's parent is not included, the node will be included as if it is a top level node
     def arrange_nodes(nodes)
-      arranged = ActiveSupport::OrderedHash.new
-      min_depth = Float::INFINITY
-      index = Hash.new { |h, k| h[k] = ActiveSupport::OrderedHash.new }
-
-      nodes.each do |node|
-        children = index[node.id]
-        index[node.parent_id][node] = children
-
-        depth = node.depth
-        if depth < min_depth
-          min_depth = depth
-          arranged.clear
-        end
-        arranged[node] = children if depth == min_depth
-      end
+      node_ids = Set.new(nodes.map(&:id))
 
-      arranged
+      index = Hash.new { |h, k| h[k] = {} }
+      nodes.each_with_object({}) do |node, arranged|
+        index[node.parent_id][node] = children = index[node.id]
+        arranged[node] = children unless node_ids.include?(node.parent_id)
```

Comment:
If I'm reading this correctly, you know all of the node ids ahead of time, and so, if it is a child it won't end up in arranged, but if it's a parent it will?  If so, that would work really well.

Reply:
Can this code be pulled out into a separate PR?  I think this method alone solves the main concern in #363, and the rest is a separate refactoring.  Additionally, I am +1 with this section because I understand it, and would merge now, but the rest I don't grasp yet.

Reply:
The change to this method ensures that all nodes are returned.

If you note, the previous tests had a bunch of `if` in them. This was because the output was non-deterministic. It turns out the code was munging the ordering sql in `arrange` and the order of the tree nodes in `sort_by_ancestry` (specifically when parent nodes were missing)

So these other changes were necessary to ensure a more consistent result.

And in the end, it was only the custom sorting with partial nodes that caused a problem (because not all nodes that needed to be sorted were present. Namely the parent)

I feel all changes are necessary fully resolve this issue.
But if you notice, I did split this up into commits similar along the lines you mentioned. Do you really want to pull the first 2 commits out as a separate PR?

Reply:
I thought they were independent, but if not, it's fine together.

Reply:
ok, they were pulled out.
You are right. they are independent, so they belong on their own.